### DISCUSSION TITLE
[v5] How do I skip rule validation when value is optional?

### DISCUSSION THREAD

**Discussion by dessonchan** [adonisjs/core#2234] [Help]
I have some code like this, when the website is null validator is giving a result of url validation filed. How can I skip the rule checking if the input value is null?
```
        const validationSchema = schema.create({
            enable: schema.boolean(),
            avatar: schema.file.optional({
                extnames: ['png', 'jpg', 'jpeg'],
            }),
            name: schema.string(),
            description: schema.string.optional(),
            website: schema.string.optional({}, [
                rules.url()
            ]),
            ...
        })
```

**Comment by 42pe** [comment]
@dessonchan why don't you create a second schema with that particular validation, remove it from the first one, and run the validator on that second schema ONLY if the web field is present on the request?

Or conditionally set the object going into `schema.create`, so it contains (or not) the `website` key.

**Reply by dessonchan** [reply]
I think this is final final workaround and this make the code not clean enough. And I do think this is a bug on validator...

**Reply by 42pe** [reply]
I assume you could also make a custom rule, something like `rules.urlOrNull()`, I need to solve a similar problem with dates, but have not gotten to it yet.

**Reply by thetutlage** [reply]
`null` is actually a value and the `optional` variant ignores only when the value is undefined. However, in the next release, we will be adding support to skip validation on null also using `string.canBeNull()` method

**Reply by 42pe** [reply]
@thetutlage out of curiosity, with "canBeNull" in the horizon, what will you be doing with to "requiredIfExists"? Will you be adding a "requiredIfNotNull"?

**Reply by dessonchan** [reply]
> `null` is actually a value and the `optional` variant ignores only when the value is undefined. However, in the next release, we will be adding support to skip validation on null also using `string.canBeNull()` method

It sounds great, Thank you for all the good job.

**Comment by thetutlage** [comment]
I just gave this code a try and it seems to be working fine for me even if I send `null` values. @dessonchan Is it possible for you to share a repo that can help me reproduce the issue?

**Comment by paulophdev** [comment]
I had a similar problem and I solved it this way:

`corporateName: schema.string.optional()`

**Comment by jcjp** [comment]
Encountering this as well even tried to use `nullableAndOptional()` on `schema.array` but still throws an error of array validation failed.

**Comment by matiasjs** [comment]
I'm using schema.enum.optional(...)
